### DISCUSSION TITLE
Can I remove prometheus exemplar labels?

### DISCUSSION THREAD

**Discussion by myeongseok-rpls** [open-telemetry/opentelemetry-java-instrumentation#12411] [Q&A] [answered]
I'm not good at English, so please understand if it's awkward.

I ran the agent using java auto-instrument.
Everything seemed to be working fine, but I didn't see any exemplar in grafana.
I checked the error below through the prometheus log.

**err="label length for exemplar exceeds maximum of 128 UTF-8 characters"**

When I checked the metric, there were too many labels attached to the exemplar.

**# client_address="127.0.0.6",network_peer_address="127.0.0.6",network_peer_port="55579",server_address="10.3.17.168",server_port="8081",span_id="049178b29912fdb4",trace_id="0669315b30dbe08683c19ed9bd24068b",url_path="/actuator/health",user_agent_original="kube-probe/1.29+"}**

Is there a way to remove the Exemplar's Label?

**Comment by laurit** [comment]
cc @jack-berg

**Reply by trask** [reply]
@open-telemetry/wg-prometheus it looks like this error is coming from the prometheus side:

https://github.com/prometheus/prometheus/blob/90f7832447d2f26c181ad42fef757be68d0cb805/storage/interface.go#L43

do you know if there's anything spec'd about this from an OTel <-> Prometheus interop perspective?

**Reply by dashpole** [reply]
Exporters have flexibility here.  See https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md#exemplars-1.

> Prometheus Exemplars MUST use the trace_id and span_id keys for the trace and span IDs, respectively. Timestamps MUST be added as timestamps on the Prometheus exemplar, and filtered_attributes MUST be added as labels on the Prometheus exemplar unless they would exceed the [limit on characters](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#exemplars).

So you MUST include the trace_id and span_id keys on the prometheus exemplar, but if filtered attributes would cause it to exceed the length limit, you are not required to add them.  A simple strategy would be to drop all filtered attributes if the total length would be longer than the limit.

**Reply by jack-berg** [reply]
Sorry I don't believe there is any way to remove or modify exemplar attributes with SDK configuration, though arguably there should be. Is it possible to configure prometheus to allow more characters?

**Reply by jack-berg** [reply]
> Prometheus Exemplars MUST use the trace_id and span_id keys for the trace and span IDs, respectively. Timestamps MUST be added as timestamps on the Prometheus exemplar, and filtered_attributes MUST be added as labels on the Prometheus exemplar unless they would exceed the [limit on characters](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#exemplars).

This portion of the spec seems incomplete. There's no way for exporters to be aware of the limit on characters.

**Reply by dashpole** [reply]
> There's no way for exporters to be aware of the limit on characters.

Worst case, the prom exporter could count the characters and compare to the limit, right?

**Reply by jack-berg** [reply]
How does the prom exporter know about the limit? Is it somehow configured by the user or included as a request parameter by the client?

**Reply by dashpole** [reply]
It is a global constant right now, at least in Go: https://github.com/prometheus/client_golang/blob/93c851f0b33308b6d31bb36d6072f44d0d839c2c/prometheus/value.go#L240

**Reply by jack-berg** [reply]
Interesting.. ok yeah limiting to a hardcoded value seems good for now. If it ever becomes configurable for prometheus, we can add corresponding configuration options in the exporter.

Ideally the spec's prometheus exporter doc would give this guidance so maintainers can be aware of this common pitfall.

**Reply by jack-berg** [reply, marked as answer]
Issue to track: https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/12417
